src/v/cluster/leader_router.h

```diff
@@ -0,0 +1,266 @@
+        }
+        auto leader = leader_opt.value();
+
+        if (leader == _self) {
+            r = co_await find_shard_and_process(req, ntp, timeout);
+        } else {
+            vlog(
+              clusterlog.trace,
+              "dispatching {} from {} to {} ",
+              handler_t::process_name(),
+              _self,
+              leader);
+            r = co_await dispatch_to_leader(req, leader, timeout);
+        }
+
+        if (likely(r.ec == errc::success)) {
+            error = std::nullopt;
+            break;
+        }
+
+        if (likely(r.ec != errc::replication_error)) {
+            error = vformat(
+              fmt::runtime("{} failed with {}"),
+              handler_t::process_name(),
+              r.ec);
+            break;
+        }
+
+        error = vformat(
+          fmt::runtime("{} failed with {}"), handler_t::process_name(), r.ec);
+        vlog(
+          clusterlog.trace,
+          "{} failed, retries left: {}",
+          handler_t::process_name(),
+          retries);
+        co_await sleep_abortable(delay_ms, _as);
+    }
+    if (error) {
+        vlog(
+          clusterlog.warn,
+          "{} failed: {}",
+          handler_t::process_name(),
+          error.value());
+    }
+
+    co_return r;
+}
+
+template<typename req_t, typename resp_t, typename handler_t>
+requires IsHandler<req_t, resp_t, handler_t>
+ss::future<resp_t> leader_router<req_t, resp_t, handler_t>::dispatch_to_leader(
+  req_t req, model::node_id leader_id, model::timeout_clock::duration timeout) {
+    using proto_t = handler_t::proto_t;
+    return _connection_cache.local()
+      .with_node_client<proto_t>(
+        _self,
+        ss::this_shard_id(),
+        leader_id,
+        timeout,
+        [req = std::move(req), timeout](proto_t proto) {
+            return handler_t::dispatch(
+              std::move(proto), std::move(req), timeout);
+        })
+      .then(&rpc::get_ctx_data<resp_t>)
+      .then([](result<resp_t> r) {
+          if (r.has_error()) {
+              vlog(
+                clusterlog.warn,
+                "got error {} on remote {}",
+                r.error().message(),
+                handler_t::process_name());
+              return handler_t::error_resp(errc::timeout);
+          }
+          return r.value();
+      });
+}
+
+template<typename req_t, typename resp_t, typename handler_t>
+requires IsHandler<req_t, resp_t, handler_t>
+ss::future<resp_t>
+leader_router<req_t, resp_t, handler_t>::find_shard_and_process(
+  req_t req, model::ntp ntp, model::timeout_clock::duration timeout) {
+    auto shard = _shard_table.local().shard_for(ntp);
+
+    auto holder = ss::gate::holder(_gate);
+    if (unlikely(!shard)) {
+        auto retries = _retries;
+        auto delay_ms = _retry_delay_ms;
+        while (!_as.abort_requested() && !shard && 0 < retries--) {
+            try {
+                co_await sleep_abortable(delay_ms, _as);
+            } catch (ss::sleep_aborted& e) {
```

Comment:
does `e` have an unused variable warning?

Reply:
Oh interesting, good catch, it doesn't show up in vim

Reply:
Done